labs/KeyWrapping/KeyWrapping.md

```diff
@@ -0,0 +1,1015 @@
+  `S`, a computation of the authentication bit to assign to `FAIL`,
+  and finally the appropriate plaintext.
+
+Notice that `FAIL` indicates a *failure* to authenticate so should be
+`False` for authenticated decryptions and `True` for failures to
+authenticate.
+
+*Hint:* Review the Cryptol primitives `head` and `tail`.
+
+```
+KWAD :
+    {n}
+    (fin n, 2 <= n, n < 2^^54) =>
+    ([128] -> [128]) -> [n+1][64] -> (Bit, [n][64])
+KWAD CIPHk' C = (FAIL, P)
+  where
+    S = zero : [n+1][64]
+    FAIL = False
+    P = zero
+```
+
+When you have successfully defined this function, you can test your
+work by `:prove`ing that `KWAE` and `KWAD` are inverses (well, at
+least for a dummy CIPHk and P of length 3 semiblocks) using the
+`KWAEInvProp`. You can also check your work against six test vectors
+by using the property `KWADTests` (this is defined later on
+in this document).
+
+```
+property KWAEInvProp S = 
+    KWAD`{3} (\a -> a-1) (KWAE (\a -> a+1) S) == (False, S)
+```
+
+# Formal Specifications of `TKW-AE` and `TKW-AD`
+
+We're briefly skipping over the definitions of `KWP-AE` and `KWP-AD`
+in Section 6.3 because the definitions of `TWK-AE` and `TWK-AD` in
+Section 7 are very similar to `KW-AE` and `KW-AD`. We'll come back to
+`KWP` a little later.
+
+**EXERCISE**: Try your hand at writing the specification for `TKW-AE`
+  and `TKW-AD` from Section 7. These functions are very similar to
+  `KW-AE` and `KW-AD` but they use the 64-bit block cipher `TDEA`
+  (also known as
+  [Triple-DES](https://en.wikipedia.org/wiki/Triple_DES)). We
+  recommend following the same steps from above and defining the helper
+  functions `TWStep`, `TWStep'`, `TW`, and `TW'` before you attempt
+  `TKW-AE` and `TKW-AD`. Test vectors are available in the
+  [kwtestvectors directory](kwtestvectors).
+
+There are some minor modifications to be made, but things should go
+easily using `KW-AE` and `KW-AD` as a reference. This is a good
+opportunity to go back through and take a close look at the type
+parameters and conditions and be sure you understand what they mean
+and how to use them. One important thing to note is that had we
+defined our functions above using a `semigroup` type parameter (rather
+than hard-code 64), our work defining the TKW family of functions
+would already be done!
+
+One important difference in the TKW family is you will use the
+Triple-DES algorithm that's implemented in the `TDEA` module we imported
+earlier. You can check the type of `TDEA` in the interpreter via `:t
+TDEA::blockEncrypt`. This has a slightly different interface than the
+block cipher we used from the `AES` module earlier.
+
+You can define a keyed `TDEA` instance as follows:
+
+```
+[K1, K2, K3] = [0x0123456789ABCDEF, 0x23456789ABCDEF01, 0x456789ABCDEF0123]
+TDEA_keyed = (\p -> TDEA::blockEncrypt (K1, K2, K3, p))
+```
+
+It is worth taking a quick look through the `TripleDES.cry` to learn a
+little bit about a particularly famous NIST test vector.
+
+Good luck!
+
```

Comment:
Doesn't look like these functions are in the answers module

Reply:
Whoops! I’ll add them ASAP!

Reply:
We also didn't add any test vectors...

Reply:
I'm going to push what I have and consider test vectors after lunch.

Reply:
Done!